### PR TITLE
WIP: Redis lettuce

### PR DESCRIPTION
**Pull request by felixbarny**
Depends on elastic/apm-agent-java#822

### Checklist

- [x] Implement code
- [ ] Add tests
- [ ] Update documentation
- [ ] Update [CHANGELOG.md](CHANGELOG.md)
- [ ] Update [supported-technologies.asciidoc](docs/supported-technologies.asciidoc)
- [ ] Added an API method or config option? Document in which version this will be introduced.

### CODE REVIEW

File: apm-agent-plugins/apm-redis-plugin/apm-lettuce-plugin/src/main/java/co/elastic/apm/agent/redis/lettuce/LettuceStopSpanInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        private static ThreadLocal<Boolean> nestedThreadLocal = new ThreadLocal<>();
+
+        @VisibleForAdvice
+        public static boolean beforeComplete(@Nullable Throwable t, boolean cancelled) {
+            try {
+                // guard against wrapped RedisCommands
+                if (nestedThreadLocal.get() == Boolean.TRUE || tracer == null) {
+                    return true;
+                }
+                TraceContextHolder<?> active = tracer.getActive();
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-redis-plugin/apm-lettuce-plugin/src/main/java/co/elastic/apm/agent/redis/lettuce/LettuceStopSpanInstrumentation.java]: So, the command methods are guaranteed to be called within the `RedisChannelWriter#write` execution? If so, can't the `RedisChannelWriter#exit` be used to end the span, without worrying about nesting (still requires the instrumentation of `completeExceptionally` for the Exception probably)? If not- relying here on `getActive` is unsafe and may cause span leak.

File: apm-agent-plugins/apm-redis-plugin/apm-lettuce-plugin/src/test/java/co/elastic/apm/agent/redis/lettuce/LettuceInstrumentationTest.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import io.lettuce.core.api.async.RedisAsyncCommands;
+import io.lettuce.core.api.reactive.RedisReactiveCommands;
+import io.lettuce.core.api.sync.RedisCommands;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class LettuceInstrumentationTest extends AbstractRedisInstrumentationTest {
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-redis-plugin/apm-lettuce-plugin/src/test/java/co/elastic/apm/agent/redis/lettuce/LettuceInstrumentationTest.java]: Can you find ways to test command cancellation and errors?

**Comment by felixbarny** [comment]
Benchmark results executed on MacBook Pro (15-inch, 2017) 2.9GHz i7

```
Benchmark              Mode instrument=false active=false active=true  Units
avg                  sample           47.923       50.083      52.942  us/op
p0.00                sample           34.880       39.872      38.400  us/op
p0.50                sample           43.136       45.440      48.192  us/op
p0.90                sample           57.408       60.480      65.088  us/op
p0.95                sample           77.562       80.128      79.488  us/op
p0.99                sample          126.976      123.904     120.448  us/op
p0.999               sample          200.448      191.488     190.720  us/op
p0.9999              sample          498.425      404.542     325.435  us/op
p1.00                sample         3559.424     1699.840    5832.704  us/op
·cpu.time.norm       sample        35415.123    39221.645   55536.996  ns/op
·gc.alloc.rate.norm  sample          624.704      625.803     628.399   B/op
```

**Comment by felixbarny** [comment]
superseded by #949

**Pull request closed without merging**